src/topology/homotopy/basic.lean

```diff
@@ -187,4 +190,86 @@ end
 
 end homotopy
 
+/--
+The type of homotopies between two functions, fixing all points in `A : set X`.
+-/
+structure homotopy_rel (f g : C(X, Y)) (A : set X) extends homotopy f g :=
+(eq_fst' : ∀ t (x ∈ A), to_fun (t, x) = f x)
+(eq_snd' : ∀ t (x ∈ A), to_fun (t, x) = g x)
+
+namespace homotopy_rel
+
+section
+
+variables {f₀ f₁ : C(X, Y)} {A : set X}
+
+instance : has_coe_to_fun (homotopy_rel f₀ f₁ A) := ⟨_, λ F, F.to_fun⟩
+
+@[simp]
+lemma coe_to_homotopy {F : homotopy_rel f₀ f₁ A} : ⇑F.to_homotopy = F := rfl
+
+lemma coe_fn_injective : @function.injective (homotopy_rel f₀ f₁ A) (I × X → Y) coe_fn :=
+begin
+  rintros ⟨⟨⟨F, _⟩, _⟩, _⟩ ⟨⟨⟨G, _⟩, _⟩, _⟩ h,
+  congr' 3,
+end
+
+@[ext]
+lemma ext {F G : homotopy_rel f₀ f₁ A} (h : ∀ t, F t = G t) : F = G :=
+coe_fn_injective $ funext h
+
+lemma eq_fst (F : homotopy_rel f₀ f₁ A) (t : I) {x : X} (hx : x ∈ A) : F (t, x) = f₀ x :=
```

Comment:
Why not use `restate_axiom` to obtain these?

Reply:
I didn't know about `restate_axiom`, but also it seems like `restate_axiom` makes `x` here explicit, but I think it can be implicit.